ci: Slim down lint image

maflcko opened the pull request:
Currently, the lint_test_runner is built and installed into the lint CI image. This is problematic, because it triggers a full image build on every change to its source code. Doing a build of the lint test_runner on every run is easier and faster.

Commit:
ci: Slim down lint image

Currently, the lint_test_runner is built and installed into the lint CI
image. This is problematic, because it triggers a full image build on
every change to its source code. Doing a build of the lint test_runner
on every run is easier and faster.

Commit:
ci: Bump lint imagefile FROM base

According to https://wiki.debian.org/DebianReleases#Production_Releases,
the current image will be EOL in about one year, so it seems fine to
slowly bump it to something more recent.

kevkevinpal commented:
utACK [fae322a](https://github.com/bitcoin/bitcoin/pull/32250/commits/fae322a43a4b71b95b68f0674a5d2bd597fee68b)

This change makes sense, it is more efficient. 
Not sure if you have run times from the build and how much faster this change makes it?

---

- Google provides a table for what is the appropriate path to upgrade and `Debian 12 "Bookworm"` -> `Ubuntu 24.04` looks right
https://cloud.google.com/software-supply-chain-security/docs/base-images#google-provided_base_images

- I also did `grep -nri "LINT_RUNNER_PATH" ./ -I` and found no results

l0rinc commented on ci/lint/04_install.sh:
```diff
@@ -12,10 +12,11 @@ pushd "/"
 # - git (used in many lint scripts)
 # - gpg (used by verify-commits)
-${CI_RETRY_EXE} apt-get install -y curl xz-utils git gpg
+${CI_RETRY_EXE} apt-get install -y cargo curl xz-utils git gpg
```
later we're reinstalling `curl` and `xz-utils`, we could remove them from there

l0rinc replied:
That's what I meant, of course, but if you think duplicating them is better for documentation, I'm fine with it of course

l0rinc commented:
Checked that we don't rebuild anymore (I trust that we did before):
```bash
DOCKER_BUILDKIT=1 docker build --file "./ci/lint_imagefile" ./ && \
echo >> ./test/lint/test_runner/src/main.rs && \
DOCKER_BUILDKIT=1 docker build -t bitcoin-linter --file "./ci/lint_imagefile" ./ && docker run --rm -v $(pwd):/bitcoin -it bitcoin-linter
```

<details>
<summary>Result</summary>

```bash
 => [1/7] FROM mirror.gcr.io/ubuntu:24.04@sha256:1e622c5f073b4f6bfad6632f2616c7f59ef256e96fe78bf6a595d1dc4376ac02                                                          0.0s
 => [internal] load build context                                                                                                                                          0.0s
 => => transferring context: 239B                                                                                                                                          0.0s
 => CACHED [2/7] COPY ./ci/retry/retry /ci_retry                                                                                                                           0.0s
 => CACHED [3/7] COPY ./.python-version /.python-version                                                                                                                   0.0s
 => CACHED [4/7] COPY --chmod=755 ./ci/lint/container-entrypoint.sh /entrypoint.sh                                                                                         0.0s
 => CACHED [5/7] COPY ./ci/lint/04_install.sh /install.sh                                                                                                                  0.0s
 => CACHED [6/7] RUN /install.sh                                                                                                                                           0.0s
 => CACHED [7/7] WORKDIR /bitcoin 
```

</details>

Left a few cleanup nits, but LGTM otherwise, ACK fae322a43a4b71b95b68f0674a5d2bd597fee68b

Commit:
ci: refactor: Use version id over version codename consistently

Also, remove the minor version from the comment, because it does not
matter and may even change at any time.

janb84 reviewed (commented):
ACK [fac9709](https://github.com/bitcoin/bitcoin/pull/32250/commits/fac97092975e75f074505c77429b57ba9bb62b3e)

- Tested ✅
   - image is slightly larger 2.82 GB vs 2.74  GB
   - image does not rebuild when the Rust source is changed ✅. (Works as advertised, also tested that the version on master rebuilds on change)
  -  (test env. Mac arm, docker desktop version 4.40)

l0rinc commented on ci/lint_imagefile:
```diff
@@ -13,7 +13,6 @@ COPY ./ci/retry/retry /ci_retry
 COPY ./.python-version /.python-version
 COPY ./ci/lint/container-entrypoint.sh /entrypoint.sh
```
It's available apparently since 2017, I would expect it to work - but haven't tried it yet, we can try it in a follow-up PR

l0rinc commented on ci/lint_imagefile:
```diff
@@ -4,7 +4,7 @@
-FROM mirror.gcr.io/debian:bookworm
+FROM mirror.gcr.io/ubuntu:24.04
 
 ENV DEBIAN_FRONTEND=noninteractive
```
What's the rationale for polluting the runtime with an environmental variable that is only needed during build (and in this case not even during build)?

l0rinc commented:
reACK fac97092975e75f074505c77429b57ba9bb62b3e

The [new links](https://github.com/bitcoin/bitcoin/compare/fae322a43a4b71b95b68f0674a5d2bd597fee68b..fac97092975e75f074505c77429b57ba9bb62b3e) also seem to point to the correct locations - though the first comment states version 13.3, while the link seems to point to version 13.2.

janb84 reviewed (commented):
Re- ACK [faeb1ba](https://github.com/bitcoin/bitcoin/commit/faeb1babe283d8d61d830ce78310ce23984543e2)

Difference since last ACK:
- small comment change.

l0rinc commented:
ACK faeb1babe283d8d61d830ce78310ce23984543e2
